docs/architecture/decisions/0010-release-and-deployment-process-for-azure-functions.md

```diff
@@ -0,0 +1,74 @@
+
+During the initial development of the APIs, we relied on the [continuous deployment feature](https://docs.microsoft.com/en-us/azure/azure-functions/functions-continuous-deployment), configured to deploy the API code from the `master` branch of the [digital-citizenship-functions](https://github.com/teamdigitale/digital-citizenship-functions) repository.
+
+This release process served us well during the initial development phase, but it has several drawbacks:
+
+1. there's no approval process that triggers the deployment of the `master` branch into production
+1. the build phase happens on the production hosts
+1. we cannot control the actual deployment process (the build and the re-load of the code happens automatically on the production hosts)
+1. Rolling back is cumbersome as it requires reverting commits on the `master` branch.
+
+## Decision
+
+We decide to structure the release and deployment process in the following steps:
+
+### 1 - Automated tests
+
+On commit to the `master` branch, the CI system checks out the commit and runs the automated tests.
+
+### 2 - Build of release artifacts
+
+On tests passing successfully, a [CI workflow](https://circleci.com/docs/2.0/workflows/) triggers the build process that generates artifacts needed for release (e.g. packs dependencies, compiles Typescript into Javascript, etc...).
+
+The generated artifact get stored in a `build` directory (e.g. a single javascript file generated by [Azure Functions Pack](https://github.com/Azure/azure-functions-pack)).
+
+### 3 - Version bump and tags
+
+On successful build, the patch version gets incremented (e.g. in `package.json`) and it gets committed along with the generated artifacts. 
+
+After commit, two git tags get created:
+
+* a `SNAPSHOT-x.x.x` tag using the bumped version
+* a `SNAPSHOT-LATEST` tag (this tag gets updated to the current commit)  
+
+Everything gets pushed to the upstream repository.
+
+### 4 - Promotion
+
+On the CI workflow, a [promotion step](https://circleci.com/docs/2.0/workflows/#holding-a-workflow-for-a-manual-approval) awaits for approval.
+
+When the promotion gets triggered, two git tags get created:
+
+* a `RELEASE-x.x.x` tag using the current version
+* a `RELEASE-LATEST` tag (this tag gets updated to the current commit)
+
+### 5 - deployment
+
+Deployment will still happen through the [continuous deployment feature](https://docs.microsoft.com/en-us/azure/azure-functions/functions-continuous-deployment) of Azure Functions. 
+
+The difference introduced with this new model is that instead of linking the continuous deployment to the `master` branch, we will link it to a git tag that we can control independently from what gets committed to the `master` branch. Also we could use a different git tag for each environment (i.e. staging and development).
+
+#### staging
+
+The staging environment will be configured for continuous deployments from the `SNAPSHOT-LATEST` tag.
```

Comment:
it's not clear if the staging environment it's a function slot and which resources (ie. cosmosdb, queues) shares with the production environment.

Reply:
in this case is a completely different infrastructure, completely separated from the production one (as I mentioned [here](https://github.com/teamdigitale/digital-citizenship/pull/17#discussion_r146107023)) we will create two infrastructures that share the same setup, changes will be first applied to staging and then to production, they will not share anything.